pkg/common/scheduler/state/state.go

```diff
@@ -0,0 +1,399 @@
+
+	return &stateBuilder{
+		ctx:               ctx,
+		logger:            logging.FromContext(ctx),
+		vpodLister:        lister,
+		capacity:          podCapacity,
+		schedulerPolicy:   schedulerPolicy,
+		nodeLister:        nodeLister,
+		statefulSetClient: kubeclient.Get(ctx).AppsV1().StatefulSets(namespace),
+		statefulSetName:   sfsname,
+		podLister:         podlister,
+		schedPolicy:       schedPolicy,
+		deschedPolicy:     deschedPolicy,
+	}
+}
+
+func (s *stateBuilder) State(reserved map[types.NamespacedName]map[string]int32) (*State, error) {
+	vpods, err := s.vpodLister()
+	if err != nil {
+		return nil, err
+	}
+
+	scale, err := s.statefulSetClient.GetScale(s.ctx, s.statefulSetName, metav1.GetOptions{})
+	if err != nil {
+		s.logger.Infow("failed to get statefulset", zap.Error(err))
+		return nil, err
+	}
+
+	free := make([]int32, 0)
+	schedulablePods := make([]int32, 0)
+	last := int32(-1)
+
+	// keep track of (vpod key, podname) pairs with existing placements
+	withPlacement := make(map[types.NamespacedName]map[string]bool)
+
+	podSpread := make(map[types.NamespacedName]map[string]int32)
+	nodeSpread := make(map[types.NamespacedName]map[string]int32)
+	zoneSpread := make(map[types.NamespacedName]map[string]int32)
+
+	noexec := &v1.Taint{
+		Key:    "node.kubernetes.io/unreachable",
+		Effect: v1.TaintEffectNoExecute,
+	}
+
+	nosched := &v1.Taint{
+		Key:    "node.kubernetes.io/unreachable",
+		Effect: v1.TaintEffectNoSchedule,
+	}
+
+	//Build the node to zone map
+	nodes, err := s.nodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	nodeToZoneMap := make(map[string]string)
+	zoneMap := make(map[string]struct{})
+	for i := 0; i < len(nodes); i++ {
+		node := nodes[i]
+		if node.Spec.Unschedulable {
+			continue //ignore node that is currently unschedulable
+		}
+
+		taints := node.Spec.Taints
+		if contains(taints, noexec) || contains(taints, nosched) {
+			continue //ignore node that is currently unschedulable
+		}
+
+		zoneName, ok := node.GetLabels()[scheduler.ZoneLabel]
+		if !ok {
+			continue //ignore node that doesn't have zone info (maybe a test setup or control node)
+		}
+
+		nodeToZoneMap[node.Name] = zoneName
+		zoneMap[zoneName] = struct{}{}
+	}
+
+	for podId := int32(0); podId < scale.Spec.Replicas && s.podLister != nil; podId++ {
+		var pod *v1.Pod
+		Retry(func() error {
+			pod, err = s.podLister.Get(PodNameFromOrdinal(s.statefulSetName, podId))
+			if err != nil {
+				// This error will result in a retry
+				return err
+			}
+			return nil
+		})
+
+		var unschedulable bool
+		annotVal, ok := pod.ObjectMeta.Annotations[constants.PodAnnotationKey] //Pod is marked for eviction - CANNOT SCHEDULE VREPS on this pod
+		if ok {
+			unschedulable, _ = strconv.ParseBool(annotVal)
+		}
+
+		nodeName := pod.Spec.NodeName
+		node, err := s.nodeLister.Get(nodeName) //Node could be marked as Unschedulable - CANNOT SCHEDULE VREPS on a pod running on this node
```

Comment:
same here. Maybe we just don't care about pod eviction. This has to be dealt somewhere else anyway

Reply:
Each plugin must be given some info to prevent it from selecting a pod that is in the process of eviction. Not sure yet of how to do that without the state collector telling it so...will think about this a bit more.